### PR TITLE
[Rust] wasmedge-sys v0.7.0

### PR DESCRIPTION
**Pull request by apepkuss**
In this PR, `wasmedge-sys` migrates to support the latest version of the C-APIs in `wasmedge v0.10.0 preview`. The version of `wasmedge-sys` comes to `v0.7.0`.

### CODE REVIEW

File: bindings/rust/wasmedge-sys/src/statistics.rs
Comment (hydai) [inline bindings/rust/wasmedge-sys/src/statistics.rs:87]: This is clear statistics, not all data in the store?

File: bindings/rust/wasmedge-sys/src/utils.rs
Comment (hydai) [inline bindings/rust/wasmedge-sys/src/utils.rs:74]: `check` is very confusing.

**Review by hydai — CHANGES_REQUESTED** [review]

File: bindings/rust/wasmedge-sys/src/loader.rs
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/loader.rs:86]: Accepted. Thanks!

File: bindings/rust/wasmedge-sys/src/statistics.rs
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/statistics.rs:87]: Accepted. Thanks!

File: bindings/rust/wasmedge-sys/src/utils.rs
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/utils.rs:74]: Accepted. Will refine the doc comment. Thanks!

**Comment by apepkuss** [comment]
Hi @hydai @q82419 ,

The CI checks reported two failures that could indicate some potential issues. Please help check the issues further.

- **binding-rust/MacOS (macos-11)** reported that the rust example `func_call_across_different_modes` failed ONLY on macOS. The issue was reproduced on my MBP (v12.0.1 + Intel). 

    The design of this example is derived from the WasmEdge test case [mixcall](https://github.com/WasmEdge/WasmEdge/tree/master/test/mixcall). Therefore, I tried the `mixcall` test case on the same MBP, and it also failed. Besides, there are also additional 4 failed test cases. The relevant test log of these failed test cases is attached below:

    - 1:wasmedgeAOTCoreTests
      ```text
      1/21 Testing: wasmedgeAOTCoreTests
      1/21 Test: wasmedgeAOTCoreTests
      Command: "/Users/xinliu/workspace/secondstate/WasmEdge/build/test/aot/wasmedgeAOTCoreTests"
      Directory: /Users/xinliu/workspace/secondstate/WasmEdge/build/test/aot
      "wasmedgeAOTCoreTests" start time: Apr 21 23:01 CST
      Output:
      ----------------------------------------------------------
      [==========] Running 316 tests from 4 test suites.
      [----------] Global test environment set-up.
      [----------] 2 tests from AsyncRunWsmFile
      [ RUN      ] AsyncRunWsmFile.NativeInterruptTest
      [2022-04-21 23:01:08.816] [error] wasmedge runtime failed: execution interrupted, Code: 0x07
      [2022-04-21 23:01:08.817] [error]     When executing function name: "_start"
      [2022-04-21 23:01:08.818] [error] wasmedge runtime failed: execution interrupted, Code: 0x07
      [2022-04-21 23:01:08.818] [error]     When executing function name: "_start"
      [       OK ] AsyncRunWsmFile.NativeInterruptTest (360 ms)
      [ RUN      ] AsyncRunWsmFile.CustomWasmInterruptTest
      <end of output>
      Test time =   2.36 sec
      ----------------------------------------------------------
      Test Failed.
      "wasmedgeAOTCoreTests" end time: Apr 21 23:01 CST
      "wasmedgeAOTCoreTests" time elapsed: 00:00:02
      ----------------------------------------------------------
      ```
    - 4:wasmedgeMixcallTests
      ```text
      4/21 Testing: wasmedgeMixcallTests
      4/21 Test: wasmedgeMixcallTests
      Command: "/Users/xinliu/workspace/secondstate/WasmEdge/build/test/mixcall/wasmedgeMixcallTests"
      Directory: /Users/xinliu/workspace/secondstate/WasmEdge/build/test/mixcall
      "wasmedgeMixcallTests" start time: Apr 21 23:01 CST
      Output:
      ----------------------------------------------------------
      [==========] Running 2 tests from 1 test suite.
      [----------] Global test environment set-up.
      [----------] 2 tests from MixCallTest
      [ RUN      ] MixCallTest.Call__InterpCallAOT
      [2022-04-21 23:01:10.029] [info] compile start
      [2022-04-21 23:01:10.031] [info] verify start
      [2022-04-21 23:01:10.031] [info] optimize start
      [2022-04-21 23:01:10.036] [info] codegen start
      [2022-04-21 23:01:10.041] [info] output start
      [2022-04-21 23:01:10.054] [info] compile done
      [2022-04-21 23:01:10.075] [info] output start
      <end of output>
      Test time =   0.44 sec
      ----------------------------------------------------------
      Test Failed.
      "wasmedgeMixcallTests" end time: Apr 21 23:01 CST
      "wasmedgeMixcallTests" time elapsed: 00:00:00
      ----------------------------------------------------------
      ```
    - 14:wasmedgeProcessTests
      ```text
      14/21 Testing: wasmedgeProcessTests
      14/21 Test: wasmedgeProcessTests
      Command: "/Users/xinliu/workspace/secondstate/WasmEdge/build/test/host/wasmedge_process/wasmedgeProcessTests"
      Directory: /Users/xinliu/workspace/secondstate/WasmEdge/build/test/host/wasmedge_process
      "wasmedgeProcessTests" start time: Apr 21 23:13 CST
      Output:
      ----------------------------------------------------------
      [==========] Running 10 tests from 1 test suite.
      [----------] Global test environment set-up.
      [----------] 10 tests from WasmEdgeProcessTest
      [ RUN      ] WasmEdgeProcessTest.SetProgName
      [       OK ] WasmEdgeProcessTest.SetProgName (0 ms)
      [ RUN      ] WasmEdgeProcessTest.AddArg
      [       OK ] WasmEdgeProcessTest.AddArg (0 ms)
      [ RUN      ] WasmEdgeProcessTest.AddEnv
      [       OK ] WasmEdgeProcessTest.AddEnv (0 ms)
      [ RUN      ] WasmEdgeProcessTest.AddStdIn
      [       OK ] WasmEdgeProcessTest.AddStdIn (0 ms)
      [ RUN      ] WasmEdgeProcessTest.SetTimeOut
      [       OK ] WasmEdgeProcessTest.SetTimeOut (0 ms)
      [ RUN      ] WasmEdgeProcessTest.Run
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/wasmedge_process/wasmedge_process.cpp:176: Failure
      Expected equality of these values:
        RetVal[0].get<int32_t>()
          Which is: -1
        1
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/wasmedge_process/wasmedge_process.cpp:177: Failure
      Value of: Env.StdOut.size() == 0
        Actual: false
      Expected: true
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/wasmedge_process/wasmedge_process.cpp:178: Failure
      Value of: Env.StdErr.size() > 0
        Actual: false
      Expected: true
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/wasmedge_process/wasmedge_process.cpp:184: Failure
      Expected equality of these values:
        RetVal[0].get<int32_t>()
          Which is: -1
        1
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/wasmedge_process/wasmedge_process.cpp:185: Failure
      Value of: Env.StdOut.size() == 0
        Actual: false
      Expected: true
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/wasmedge_process/wasmedge_process.cpp:186: Failure
      Value of: Env.StdErr.size() > 0
        Actual: false
      Expected: true
      [  FAILED  ] WasmEdgeProcessTest.Run (10 ms)
      [ RUN      ] WasmEdgeProcessTest.GetExitCode
      [       OK ] WasmEdgeProcessTest.GetExitCode (0 ms)
      [ RUN      ] WasmEdgeProcessTest.GetStdOut
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/wasmedge_process/wasmedge_process.cpp:224: Failure
      Expected equality of these values:
        RetVal[0].get<uint32_t>()
          Which is: 4294967295
        0U
          Which is: 0
      [  FAILED  ] WasmEdgeProcessTest.GetStdOut (1 ms)
      [ RUN      ] WasmEdgeProcessTest.GetStdErr
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/wasmedge_process/wasmedge_process.cpp:252: Failure
      Value of: Len > 0
        Actual: false
      Expected: true
      [  FAILED  ] WasmEdgeProcessTest.GetStdErr (1 ms)
      [ RUN      ] WasmEdgeProcessTest.Module
      [       OK ] WasmEdgeProcessTest.Module (0 ms)
      [----------] 10 tests from WasmEdgeProcessTest (13 ms total)
      
      [----------] Global test environment tear-down
      [==========] 10 tests from 1 test suite ran. (14 ms total)
      [  PASSED  ] 7 tests.
      [  FAILED  ] 3 tests, listed below:
      [  FAILED  ] WasmEdgeProcessTest.Run
      [  FAILED  ] WasmEdgeProcessTest.GetStdOut
      [  FAILED  ] WasmEdgeProcessTest.GetStdErr
      
       3 FAILED TESTS
      <end of output>
      Test time =   0.34 sec
      ----------------------------------------------------------
      Test Failed.
      "wasmedgeProcessTests" end time: Apr 21 23:13 CST
      "wasmedgeProcessTests" time elapsed: 00:00:00
      ----------------------------------------------------------
      ```
    - 15:wasiSocketTests
      ```text
      15/21 Testing: wasiSocketTests
      15/21 Test: wasiSocketTests
      Command: "/Users/xinliu/workspace/secondstate/WasmEdge/build/test/host/socket/wasiSocketTests"
      Directory: /Users/xinliu/workspace/secondstate/WasmEdge/build/test/host/socket
      "wasiSocketTests" start time: Apr 21 23:13 CST
      Output:
      ----------------------------------------------------------
      [==========] Running 2 tests from 1 test suite.
      [----------] Global test environment set-up.
      [----------] 2 tests from WasiTest
      [ RUN      ] WasiTest.SocketUDP
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/socket/wasi_socket.cpp:106: Failure
      Expected equality of these values:
        Errno[0].get<int32_t>()
          Which is: 67
        __WASI_ERRNO_SUCCESS
          Which is: 0
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/socket/wasi_socket.cpp:115: Failure
      Expected equality of these values:
        Errno[0].get<int32_t>()
          Which is: 67
        __WASI_ERRNO_SUCCESS
          Which is: 0
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/socket/wasi_socket.cpp:138: Failure
      Expected equality of these values:
        Errno[0].get<int32_t>()
          Which is: 8
        __WASI_ERRNO_SUCCESS
          Which is: 0
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/socket/wasi_socket.cpp:160: Failure
      Expected equality of these values:
        Errno[0].get<int32_t>()
          Which is: 8
        __WASI_ERRNO_SUCCESS
          Which is: 0
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/socket/wasi_socket.cpp:177: Failure
      Expected equality of these values:
        Errno[0].get<int32_t>()
          Which is: 8
        __WASI_ERRNO_SUCCESS
          Which is: 0
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/socket/wasi_socket.cpp:180: Failure
      Expected equality of these values:
        MsgRecv
          Which is: "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"
        Msg1
          Which is: "hello, wasmedge."
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/socket/wasi_socket.cpp:184: Failure
      Expected equality of these values:
        Errno[0].get<int32_t>()
          Which is: 8
        __WASI_ERRNO_SUCCESS
          Which is: 0
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/socket/wasi_socket.cpp:187: Failure
      Expected equality of these values:
        Errno[0].get<int32_t>()
          Which is: 8
        __WASI_ERRNO_SUCCESS
          Which is: 0
      [  FAILED  ] WasiTest.SocketUDP (0 ms)
      [ RUN      ] WasiTest.GetAddrinfo
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/socket/wasi_socket.cpp:375: Failure
      Expected equality of these values:
        Errno[0].get<int32_t>()
          Which is: 0
        __WASI_ERRNO_AIBADFLAG
          Which is: 79
      [  FAILED  ] WasiTest.GetAddrinfo (75 ms)
      [----------] 2 tests from WasiTest (75 ms total)
      
      [----------] Global test environment tear-down
      [==========] 2 tests from 1 test suite ran. (75 ms total)
      [  PASSED  ] 0 tests.
      [  FAILED  ] 2 tests, listed below:
      [  FAILED  ] WasiTest.SocketUDP
      [  FAILED  ] WasiTest.GetAddrinfo
      
       2 FAILED TESTS
      <end of output>
      Test time =   0.39 sec
      ----------------------------------------------------------
      Test Failed.
      "wasiSocketTests" end time: Apr 21 23:13 CST
      "wasiSocketTests" time elapsed: 00:00:00
      ----------------------------------------------------------
      ```
    - 16:wasiTests 
      ```text
      16/21 Testing: wasiTests
      16/21 Test: wasiTests
      Command: "/Users/xinliu/workspace/secondstate/WasmEdge/build/test/host/wasi/wasiTests"
      Directory: /Users/xinliu/workspace/secondstate/WasmEdge/build/test/host/wasi
      "wasiTests" start time: Apr 21 23:13 CST
      Output:
      ----------------------------------------------------------
      [==========] Running 9 tests from 1 test suite.
      [----------] Global test environment set-up.
      [----------] 9 tests from WasiTest
      [ RUN      ] WasiTest.Args
      [       OK ] WasiTest.Args (0 ms)
      [ RUN      ] WasiTest.Envs
      [       OK ] WasiTest.Envs (0 ms)
      [ RUN      ] WasiTest.ClockRes
      [       OK ] WasiTest.ClockRes (0 ms)
      [ RUN      ] WasiTest.PollOneoffSocket
      /Users/xinliu/workspace/secondstate/WasmEdge/test/host/wasi/wasi.cpp:655: Failure
      Expected equality of these values:
        Errno[0].get<int32_t>()
          Which is: 52
        __WASI_ERRNO_SUCCESS
          Which is: 0
      [  FAILED  ] WasiTest.PollOneoffSocket (636 ms)
      [ RUN      ] WasiTest.ClockTimeGet
      [       OK ] WasiTest.ClockTimeGet (0 ms)
      [ RUN      ] WasiTest.ProcExit
      [       OK ] WasiTest.ProcExit (0 ms)
      [ RUN      ] WasiTest.Random
      [       OK ] WasiTest.Random (0 ms)
      [ RUN      ] WasiTest.Directory
      [       OK ] WasiTest.Directory (0 ms)
      [ RUN      ] WasiTest.SymbolicLink
      [       OK ] WasiTest.SymbolicLink (0 ms)
      [----------] 9 tests from WasiTest (638 ms total)
      
      [----------] Global test environment tear-down
      [==========] 9 tests from 1 test suite ran. (638 ms total)
      [  PASSED  ] 8 tests.
      [  FAILED  ] 1 test, listed below:
      [  FAILED  ] WasiTest.PollOneoffSocket
      
       1 FAILED TEST
      <end of output>
      Test time =   0.96 sec
      ----------------------------------------------------------
      Test Failed.
      "wasiTests" end time: Apr 21 23:13 CST
      "wasiTests" time elapsed: 00:00:00
      ----------------------------------------------------------
      ```

- **Build and Test WasmEdge on multiple platforms / Windows 10** reported a failure that should be caused by failing to download "winsdk_19041.iso". The detailed log info can be found [here](https://github.com/WasmEdge/WasmEdge/runs/6115871255?check_suite_focus=true#step:4:9).

File: bindings/rust/wasmedge-sys/src/utils.rs
Comment (hydai) [reply bindings/rust/wasmedge-sys/src/utils.rs:74]: I am not sure `check` is meaningful. This function seems like that trying to convert `WasmEdge_Result` to `WasmEdgeResult`.

Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/utils.rs:74]: From the perspective of the internal implementation of this function, yes, you're right, it actually converts the raw `WasmEdge_Result` into the result type of `wasmedge-sys`. But, consider the following use of the function in the practice. The semantics of `check` could be better for a user to understand the code, while `convert` could result in some kind of "misleading". Hence, in this view, using "check" should be better. Thanks!

```rust
let path = utils::path_to_cstring(path.as_ref())?;
let mod_name: WasmEdgeString = mod_name.as_ref().into();
unsafe {
    check(ffi::WasmEdge_VMRegisterModuleFromFile(self.inner.0, mod_name.as_raw(), path.as_ptr()))?
};

Ok(())
```

**Comment by hydai** [comment]
Hi @apepkuss 
After we applied the Plug-In system, some function signatures of the WasmEdge process module are changed. Could you please update it to fix the current compilation failure. Thanks

**Comment by apepkuss** [comment]
> Hi @apepkuss After we applied the Plug-In system, some function signatures of the WasmEdge process module are changed. Could you please update it to fix the current compilation failure. Thanks

@hydai Sure! Where can I get the latest version of WasmEdge? master branch or some PR? I'll refactor the affected parts of wasmedge-sys based on it.

**Comment by hydai** [comment]
Please base on master branch, which will be the tag 0.10.0-alpha.1 soon.

**Comment by apepkuss** [comment]
@hydai 

The code in this PR has already been refactored to support the updated `WasmEdge_ModuleInstanceInitWasmEdgeProcess` function and the new `WasmEdge_Plugin_loadWithDefaultPluginPaths` function. In addition, the CI workflow of `bindings-rust` has also been updated. Please help review the changes. Thanks a lot!

**Comment by apepkuss** [comment]
@hydai 

There is still a failure on macOS, so please delay the review until I fixed it. Thanks!

**Comment by apepkuss** [comment]
@hydai 

The issues in `build.rs`, which are related to the path to the `wasmedge_process` plugin on macOS, have been fixed and tested. According to the latest CI checks, there is a test case that failed on macOS. I tried to reproduce the failure in my local environment (macOS 12.3.1 + Apple Silicon) but did not observe the failure again; in addition, the latest changes for supporting the new two APIs have no effects on this test case. Based on these reasons, I suggest that we skip the failure just in this PR and try to fix it in the future. Thanks!

File: bindings/rust/wasmedge-sys/src/executor.rs
Comment (hydai) [inline bindings/rust/wasmedge-sys/src/executor.rs:108]: Do we need to keep this part?

**Review by hydai — CHANGES_REQUESTED** [review]
The last step before merge it :-)

File: bindings/rust/wasmedge-sys/src/executor.rs
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/executor.rs:108]: Accepted. Thanks a lot!

**Review by hydai — APPROVED** [review]

**Comment by hydai** [comment]
LGTM. Will merge it after the CI passes. Thanks @apepkuss

**Comment by apepkuss** [comment]
@hydai Thanks for the review.

**Pull request merged**
